### PR TITLE
fix(memcache): preserve float values in Redis-based cache backends

### PR DESCRIPTION
**Pull request by joshtrichards**
* Resolves: #  (hard to say)

## Summary

Preserve float values in Redis/Valkey-backed memcache backends.

This updates `Redis` and `KeyValueCache` to:
- decode values via `json_decode(...)` instead of coercing all numeric payloads to `int`
- encode floats with `JSON_PRESERVE_ZERO_FRACTION` so `1.0` stays distinct from `1`

It also adds shared round-trip test coverage in `tests/lib/Memcache/Cache.php`, including floats and numeric edge cases.

### Before
- `3.14` → `3`
- `1.0` → `1`

### After
- `3.14` → `3.14`
- `1.0` → `1.0`

## Notes

This slightly changes the cache format for values like `1.0`, so older entries may not compare equal in Redis-side CAS/Lua operations until they expire and are rewritten.

Down the road we may want to consider using `_pack` / `_unpack` (at least for phpredis; not sure about predis).

## TODO

- [ ] Drop redundant round-trip tests from `KeyValueCacheTest` once deciding what to do about the embedded comments there

## Checklist

- Code is [properly formatted](https://docs.nextcloud.com/server/latest/developer_manual/digging_deeper/continuous_integration.html#linting)
- [Sign-off message](https://github.com/src-d/guide/blob/master/developer-community/fix-DCO.md) is added to all commits
- [x] Tests ([unit](https://docs.nextcloud.com/server/latest/developer_manual/app_development/tutorial.html#unit-tests), [integration](https://docs.nextcloud.com/server/latest/developer_manual/app_development/tutorial.html#integration-tests), api and/or acceptance) are included
- [ ] Screenshots before/after for front-end changes
- [ ] Documentation ([manuals](https://github.com/nextcloud/documentation/) or wiki) has been updated or is not required
- [ ] [Backports requested](https://github.com/nextcloud/backportbot/#usage) where applicable (ex: critical bugfixes)
- [x] [Labels added](https://github.com/nextcloud/server/labels) where applicable (ex: bug/enhancement, `3. to review`, feature component)
- [x] [Milestone added](https://github.com/nextcloud/server/milestones) for target branch/version (ex: 32.x for `stable32`)

## AI (if applicable)

- [ ] The content of this PR was partly or fully generated using AI

### CODE REVIEW

File: lib/private/Memcache/Redis.php [hunk 1]
```diff
@@ -233,10 +233,10 @@ protected function evalLua(string $scriptName, array $keys, array $args) {
 	}
 
 	protected static function encodeValue(mixed $value): string {
-		return is_int($value) ? (string)$value : json_encode($value);
+		return is_int($value) ? (string)$value : json_encode($value, JSON_THROW_ON_ERROR | JSON_PRESERVE_ZERO_FRACTION);
 	}
 
 	protected static function decodeValue(string $value): mixed {
-		return is_numeric($value) ? (int)$value : json_decode($value, true);
+		return json_decode($value, true, 512, JSON_THROW_ON_ERROR);
```
Comment (susnux) [inline lib/private/Memcache/Redis.php:240]:
```suggestion
		return json_decode($value, true, flags: JSON_THROW_ON_ERROR);
```

File: lib/private/Memcache/KeyValueCache.php [hunk 2]
```diff
@@ -242,13 +242,13 @@ private function normalizeTtl(int $ttl): int {
...
-		return is_int($value) ? (string)$value : json_encode($value, JSON_THROW_ON_ERROR);
+		return is_int($value) ? (string)$value : json_encode($value, JSON_THROW_ON_ERROR | JSON_PRESERVE_ZERO_FRACTION);
 	}
 
 	/**
 	 * Unserialize a value from the key-value store.
 	 */
 	protected static function decodeValue(string $value): mixed {
-		return is_numeric($value) ? (int)$value : json_decode($value, true);
+		return json_decode($value, true, 512, JSON_THROW_ON_ERROR);
```
Comment (susnux) [inline lib/private/Memcache/KeyValueCache.php:252]:
```suggestion
		return json_decode($value, true, flags: JSON_THROW_ON_ERROR);
```

**Review by susnux — APPROVED** [review]
Looks good but instead of using the default you can just use the named parameter so its a bit more clear why it is set (you not actually setting the depth but only the flags)
